### PR TITLE
Replace NetworkX by RetworkX

### PR DESCRIPTION
**Pull request by maliasadi**
**Context:**
In [CircuitGraph](https://pennylane.readthedocs.io/en/stable/code/api/pennylane.CircuitGraph.html), replace [NetworkX](https://networkx.org) (NX) by [RetworkX](https://qiskit.org/documentation/retworkx) (RX) and add RX support to [QAOA](https://pennylane.readthedocs.io/en/stable/code/qml_qaoa.html) in PennyLane. 

**Description of the Change:**
RX is written in **Rust** providing a general purpose graph library. This library pursues to offer functionalities similar to what NX offers but with much faster performance. Pennylane uses NX to implement circuit graphs and QAOA with many use cases in the entire library. This PR replaces NX by RX and investigates the performance of RX in Pennylane.

**Benefits:**
Using RX instead of NX, PennyLane can take advantage of
- The performance and safety that [Rust](https://www.rust-lang.org/) provides; and 
- Faster (un)directed graph operations for [CircuitGraph](https://pennylane.readthedocs.io/en/stable/code/api/pennylane.CircuitGraph.html) and [QAOA](https://pennylane.readthedocs.io/en/stable/code/qml_qaoa.html), and their applications.

**Benchmarks:**
Check [here](https://github.com/PennyLaneAI/pennylane/pull/1791#issuecomment-998173905)

### CODE REVIEW

**Comment by maliasadi** [comment]
> @maliasadi can I ask the status of this PR? Is it mainly a prototype that can be closed, or is it still WIP?

Hey @josh146. To replace all the use cases of NetworkX with RetworkX, we need some functionalities that is planned to be added to the next release of RetworkX. So, we were waiting for the next release indeed. We can close the PR and open it again when we have every pieces of the puzzle.

File: pennylane/circuit_graph.py
Comment (josh146) [inline pennylane/circuit_graph.py:180]: How come you need to check the node doesn't exist in RX? Do you get an error otherwise?

File: pennylane/circuit_graph.py
Comment (josh146) [inline pennylane/circuit_graph.py:191]: how come the `""` is needed?

File: pennylane/circuit_graph.py
Comment (josh146) [inline pennylane/circuit_graph.py:338]: Could you explain the additional for loop here (for my own understanding!) 🙂

File: pennylane/circuit_graph.py
Comment (josh146) [inline pennylane/circuit_graph.py:673]: nice!

File: pennylane/circuit_graph.py
Comment (josh146) [inline pennylane/circuit_graph.py:682]: just double checking here, but `RX` does not have a `has_path` equivalent? The reason I ask is that we previously added this because just checking that the path exists in NX ended up being much faster than _getting_ the shortest path.

But maybe this speed difference doesn't affect RX?

File: pennylane/qaoa/cost.py
Comment (josh146) [inline pennylane/qaoa/cost.py:187]: really nice that this works for both!

File: pennylane/qaoa/cost.py
Comment (josh146) [inline pennylane/qaoa/cost.py:293]: nice docstring!

File: pennylane/qaoa/cost.py
Comment (josh146) [reply pennylane/qaoa/cost.py:187]: just so I understand, the `get_nvalue` is because `PyGraph` nodes have to be indices, and can't be objects?

Might be good to add a comment here for future developers

File: pennylane/qaoa/cost.py
Comment (josh146) [inline pennylane/qaoa/cost.py:575]:
```suggestion
        graph (nx.Graph or rx.PyGraph or rx.PyDiGraph): the directed graph on which the Hamiltonians are defined
```

File: pennylane/qaoa/cost.py
Comment (josh146) [inline pennylane/qaoa/cost.py:699]: this function did not need to be modified?

File: pennylane/qaoa/cycle.py
Comment (josh146) [inline pennylane/qaoa/cycle.py:65]:
```suggestion
        graph (nx.Graph or rx.PyGraph or rx.PyDiGraph): the graph specifying possible edges
```

File: pennylane/qaoa/cycle.py
Comment (josh146) [inline pennylane/qaoa/cycle.py:72]:
```suggestion
    if isinstance(graph, (rx.PyGraph, rx.PyDiGraph)):
```

File: pennylane/qaoa/cycle.py
Comment (josh146) [inline pennylane/qaoa/cycle.py:124]:
```suggestion
    if isinstance(graph, (rx.PyGraph, rx.PyDiGraph)):
```

File: pennylane/qaoa/cycle.py
Comment (josh146) [inline pennylane/qaoa/cycle.py:117]:
```suggestion
        graph (nx.Graph or rx.PyGraph or rx.PyDiGraph): the graph specifying possible edges
```

**Review by josh146 — CHANGES_REQUESTED** [review]
Wonderful work @maliasadi!

Very happy to approve conditional on updating the changelog to add an entry on this change.

Do you have any benchmark plots on using PL before/after this change?

File: pennylane/qaoa/cycle.py
Comment (mlxd) [inline pennylane/qaoa/cycle.py:29]: Assuming we have `from typing import Union`
we can do
```suggestion
def edges_to_wires(graph: Union[nx.Graph, rx.Graph]) -> Dict[Tuple, int]:
```

File: pennylane/qaoa/cycle.py
Comment (mlxd) [inline pennylane/qaoa/cycle.py:437]: And here.

File: pennylane/qaoa/cycle.py
Comment (mlxd) [inline pennylane/qaoa/cycle.py:574]: I assume the sort here is necessary?

**Review by mlxd — COMMENTED** [review]
Heroic effort @maliasadi ! Great work on this.
I agree with @josh146's comments, and added some small rec's in there.

File: pennylane/qaoa/cycle.py
Comment (josh146) [reply pennylane/qaoa/cycle.py:574]: nice catch, I missed this one!

**Comment by maliasadi** [comment]
Thank you @josh146  for the great suggestions! 

> Do you have any benchmark plots on using PL before/after this change?

Currently, I have benchmarks comparing NX vs. RX methods used in PL:
Create
Compute

File: pennylane/circuit_graph.py
Comment (maliasadi) [reply pennylane/circuit_graph.py:180]: This is one of those few RX and NX differences that took me hours to find out about! If you don't check `wire[0]` is already added to the graph, RX will create a new node with the same value but a different index, while NX doesn't add any new node.

File: pennylane/circuit_graph.py
Comment (maliasadi) [reply pennylane/circuit_graph.py:191]: There isn't any default value for the edge-data in RX! You can check the doc [here](https://qiskit.org/documentation/retworkx/stubs/retworkx.PyDiGraph.add_edge.html#retworkx.PyDiGraph.add_edge).

File: pennylane/circuit_graph.py
Comment (maliasadi) [reply pennylane/circuit_graph.py:338]: This is a good one! :slightly_smiling_face: the `ancestors` method in RX returns node indexes instead of node-values! So, after computing the union of ancestors of `o` in `ops`, the additional for loop fetches the node-values.

File: pennylane/circuit_graph.py
Comment (maliasadi) [reply pennylane/circuit_graph.py:682]: Yep! This is one of those missing functions in RX comparing to NX. RX `dijkstra_shortest_path` is still faster that NX `has_path` (see the benchmarks) Moreover, I will keep updating the use of RX in PL with the next releases of RX in the future.

File: pennylane/qaoa/cost.py
Comment (maliasadi) [reply pennylane/qaoa/cost.py:575]: Thanks! Fixed in all places :slightly_smiling_face:

File: pennylane/qaoa/cost.py
Comment (maliasadi) [reply pennylane/qaoa/cost.py:699]: It doesn't need any modification as I already ensured that the results of all the subroutines are exactly the same either using RX or NX (to do so, I used some hacking like sorting the returned values...)

File: pennylane/qaoa/cycle.py
Comment (maliasadi) [reply pennylane/qaoa/cycle.py:29]: Ah! Thanks a lot @mlxd for catching this. I fixed this in all places.

File: pennylane/qaoa/cycle.py
Comment (maliasadi) [reply pennylane/qaoa/cycle.py:574]: Yes indeed! One of the difficulties to get support for both RX and NX was that the results of some low-level routines aren't in the same order! Here, sorting the results of `out_edges` and `in_edges` is my trick to get equivalent results either using RX or NX.

File: doc/releases/changelog-dev.md
Comment (josh146) [inline doc/releases/changelog-dev.md:126]:
```suggestion
* The `CircuitGraph`, used to represent circuits via directed acyclic graphs, now
  uses RetworkX for its internal representation. This results in significant speedup
  for algorithms that rely on a directed acyclic graph representation.
  [(#1791)](https://github.com/PennyLaneAI/pennylane/pull/1791)
  
* The QAOA module now accepts both NetworkX and RetworkX graphs as function inputs.
  [(#1791)](https://github.com/PennyLaneAI/pennylane/pull/1791)
```

I've split this up into two entries, since they are ostensibly two separate changes (in retrospect, probably should have been two separate PRs!)

File: pennylane/circuit_graph.py
Comment (josh146) [reply pennylane/circuit_graph.py:182]: nice catch in that case! Could you add a comment on this to the code? I imagine this might be something future developers wonder about

File: pennylane/circuit_graph.py
Comment (josh146) [reply pennylane/circuit_graph.py:193]: Ah, got it! I think this is another case that would benefit from a code comment

File: pennylane/circuit_graph.py
Comment (josh146) [reply pennylane/circuit_graph.py:340]: thanks @maliasadi for explaining! as above might be good to leave a code comment

File: pennylane/circuit_graph.py
Comment (josh146) [inline pennylane/circuit_graph.py:679]: I missed this the first time around, but how come this is a private function? Is it not part of the RX public API?

**Review by josh146 — APPROVED** [review]
Thanks @maliasadi, happy to approve! 💪 

> Currently, I have benchmarks comparing NX vs. RX methods used in PL

These benchmarks look great 😍 

I have one request: do you think you would be able to quickly benchmark a more user-facing workflow? For example:

- the resulting speedup for computing cost and mixer hamiltonians in the QAOA module

- resulting speedup when computing something like the `metric_tensor`, for a circuit with a lot of layers (internally, the metric tensor computes a DAG in order to identify the 'layers' of a circuit).

More user-facing benchmarks would be super helpful, as it allows us to advertise the practical effects of this change next release 🙂

File: pennylane/circuit_graph.py
Comment (maliasadi) [reply pennylane/circuit_graph.py:679]: This is not part of the [public API](https://qiskit.org/documentation/retworkx/api.html) right now. I found it by looking into the code, but I am expecting to be added as a new feature within next releases.

**Review by mlxd — APPROVED** [review]

**Comment by maliasadi** [comment]
> * resulting speedup when computing something like the `metric_tensor`, for a circuit with a lot of layers (internally, the metric tensor computes a DAG in order to identify the 'layers' of a circuit).

I collected results for test cases in `transforms/test_metric_tensor.py` on a machine  with `Intel Core i7-1185G7 @ 3.00GHz`:
metric_tensor

Regarding QAOA routines, they are not using any advanced graph routines, and so, there shouldn't be significant speedup by using RX over NX. However, building graphs in RX is much faster and optimizer than NX (check [here](https://github.com/PennyLaneAI/pennylane/pull/1791#issuecomment-998173905) for benchmarks) and so there are big improvements on the execution-time of QAOA examples. In `./tests/test_qaoa.py`, there were 79 tests and it took ~4.15s to run, now we test 129 tests (including RX test cases) and takes ~4.33s.

File: pennylane/circuit_graph.py
Comment (maliasadi) [reply pennylane/circuit_graph.py:182]: Ooops! I thought that I already commented on this.

File: pennylane/circuit_graph.py
Comment (maliasadi) [reply pennylane/circuit_graph.py:197]: Hi @josh146. Could you please check my comments before merging? :slightly_smiling_face:

File: pennylane/circuit_graph.py
Comment (josh146) [inline pennylane/circuit_graph.py:685]: oh so there is a public equivalent?

**Review by josh146 — APPROVED** [review]
Comments look good to me @maliasadi! And thanks for the additional benchmarking :)

File: pennylane/circuit_graph.py
Comment (maliasadi) [reply pennylane/circuit_graph.py:685]: Yes! This is the public equivalent of what I used early on.

**Pull request merged**
